pkg/listwatch/listwatch.go

```diff
@@ -44,56 +50,144 @@ import (
 //
 // If the allowed namespaces includes exactly one entry with the value v1.NamespaceAll (empty string),
 // the given denied namespaces are applied.
-func NewUnprivilegedNamespaceListWatchFromClient(
+func NewNamespaceListWatchFromClient(
 	ctx context.Context,
 	l log.Logger,
-	c cache.Getter,
+	k8sVersion version.Info,
+	corev1Client corev1.CoreV1Interface,
+	ssarClient authv1.SelfSubjectAccessReviewInterface,
 	allowedNamespaces, deniedNamespaces map[string]struct{},
-) cache.ListerWatcher {
+) (cache.ListerWatcher, bool, error) {
 	if l == nil {
 		l = log.NewNopLogger()
 	}
 
-	// If the only namespace given is `v1.NamespaceAll`, then this
-	// cache.ListWatch must be privileged. In this case, return a regular
-	// cache.ListWatch tweaked with denylist fieldselector
-	// filtering the given denied namespaces.
+	listWatchAllowed, reasons, err := k8sutil.IsAllowed(
+		ctx,
+		ssarClient,
+		nil, // namespaces is a cluster-scoped resource.
+		k8sutil.ResourceAttribute{
+			Resource: "namespaces",
+			Verbs:    []string{"list", "watch"},
+		},
+	)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// The "kubernetes.io/metadata.name" label is GA since Kubernetes 1.22.
+	var metadataNameLabelSupported bool
+	v, err := semver.ParseTolerant(k8sVersion.String())
+	if err != nil {
+		level.Warn(l).Log("msg", "failed to parse Kubernetes version", "version", k8sVersion.String(), "err", err)
+	} else {
+		metadataNameLabelSupported = v.GTE(semver.MustParse("1.22.0"))
+	}
+
 	if IsAllNamespaces(allowedNamespaces) {
+		if !listWatchAllowed {
+			err := fmt.Errorf("missing list/watch permissions on the 'namespaces' resource")
+			for _, r := range reasons {
+				err = fmt.Errorf("%w: %w", err, r)
+			}
+
+			return nil, false, err
+		}
+
+		// deniedNamespaces is only supported with allowedNamespaces = "all".
 		tweak := func(options *metav1.ListOptions) {
 			DenyTweak(options, "metadata.name", deniedNamespaces)
 		}
+		if metadataNameLabelSupported {
+			// Using a label selector is more efficient but requires Kubernetes 1.22 at least.
+			tweak = func(options *metav1.ListOptions) {
+				TweakByLabel(options, "kubernetes.io/metadata.name", ExcludeFilterType, deniedNamespaces)
+			}
+		}
 
-		return cache.NewFilteredListWatchFromClient(c, "namespaces", metav1.NamespaceAll, tweak)
+		level.Debug(l).Log("msg", "using privileged namespace lister/watcher")
```

Comment:
this debug log seems out of place 🤔

Reply:
hmm not sure to understand but I can remove all the "using (un)privileged namespace lister/watcher" debug logs as they were more for me to check the code path.

Reply:
It's fine to keep it! I missed the return literally in the next line and thought you left duplicated logs 😅